pkg/controllers/proxy/network_services_controller.go

```diff
@@ -330,34 +336,74 @@ func (nsc *NetworkServicesController) Run(healthChan chan<- *healthcheck.Control
 		return errors.New("Error setting up ipvs firewall: " + err.Error())
 	}
 
-	// loop forever unitl notified to stop on stopCh
-	for {
-		select {
-		case <-stopCh:
-			glog.Info("Shutting down network services controller")
-			return nil
-		default:
-		}
+	gracefulTicker := time.NewTicker(5 * time.Second)
+	defer gracefulTicker.Stop()
 
-		glog.V(1).Info("Performing periodic sync of ipvs services")
-		err := nsc.sync()
+	select {
+	case <-stopCh:
+		glog.Info("Shutting down network services controller")
+		return nil
+	default:
+		err := nsc.doSync()
 		if err != nil {
-			glog.Errorf("Error during periodic ipvs sync in network service controller. Error: " + err.Error())
-			glog.Errorf("Skipping sending heartbeat from network service controller as periodic sync failed.")
-		} else {
-			healthcheck.SendHeartBeat(healthChan, "NSC")
+			glog.Fatalf("Failed to perform initial full sync %s", err.Error())
 		}
 		nsc.readyForUpdates = true
+	}
+
+	// loop forever until notified to stop on stopCh
+	for {
 		select {
 		case <-stopCh:
 			glog.Info("Shutting down network services controller")
 			return nil
+
+		case <-gracefulTicker.C:
+			if nsc.readyForUpdates && nsc.gracefulTermination {
+				glog.V(3).Info("Performing periodic graceful destination cleanup")
+				nsc.gracefulSync()
+			}
+
+		case perform := <-nsc.syncChan:
+			switch perform {
+			case synctypeAll:
+				glog.V(1).Info("Performing requested full sync of services")
+				err := nsc.doSync()
+				if err != nil {
+					glog.Errorf("Error during full sync in network service controller. Error: " + err.Error())
+				}
+			case synctypeIpvs:
+				glog.V(1).Info("Performing requested sync of ipvs services")
+				nsc.mu.Lock()
+				err := nsc.syncIpvsServices(nsc.serviceMap, nsc.endpointsMap)
+				nsc.mu.Unlock()
+				if err != nil {
+					glog.Errorf("Error during ipvs sync in network service controller. Error: " + err.Error())
+				}
+			}
+
 		case <-t.C:
+			glog.V(1).Info("Performing periodic sync of ipvs services")
+			err := nsc.doSync()
+			if err != nil {
+				glog.Errorf("Error during periodic ipvs sync in network service controller. Error: " + err.Error())
+				glog.Errorf("Skipping sending heartbeat from network service controller as periodic sync failed.")
+			} else {
+				healthcheck.SendHeartBeat(healthChan, "NSC")
+			}
 		}
 	}
 }
 
-func (nsc *NetworkServicesController) sync() error {
+func (nsc *NetworkServicesController) sync(syncType int) {
```

Comment:
The new behaviour is instead of having a blocking sync() on every update we have a overflowing channel, if there is a request to sync while we are syncing we will still do it but we won't have X-XX amount of them hanging in the air my the mutex in it